src/tactic/sort.lean

```diff
@@ -0,0 +1,186 @@
+| (some hyp) := get_local hyp >>= infer_type >>= recurse_on_expr t hyp rel
+| none       := target >>= recurse_on_expr t none rel
+
+end with_rel
+
+/--  The order on `polynomial.monomial n r`, where monomials are compared by their "exponent" `n`.
+If the expression is not a monomial, then the weight is `⊥`. -/
+meta def monomial_weight (e : expr) : option ℕ :=
+match e.app_fn with
+| `(coe_fn $ polynomial.monomial %%n) := n.to_nat
+| _ := none
+end
+
+/--  The function we use to compare two `expr`:
+* all non-monomials are compared alphabetically;
+* all non-monomials are smaller than all monomials;
+* bigger monomials have higher exponent.
+-/
+meta def compare_fn (eₗ eᵣ : expr) : bool :=
+match (monomial_weight eₗ, monomial_weight eᵣ) with
+| (none, none)     := eₗ.to_string ≤ eᵣ.to_string -- this solution forces an unique ordering
+| (_, none)        := false
+| (none, _)        := true
+| (some l, some r) := l ≤ r
+end
+
+/--  A version of `sort_summands_aux` that allows failure, if `allow_failure = tt`. -/
+meta def sort_summands_core (allow_failure : bool) (t : list expr) (hyp : option name) :
+  tactic unit :=
+if allow_failure then sort_summands_aux compare_fn t hyp <|> skip
+else sort_summands_aux compare_fn t hyp
+
+/-- If the target is an equality involving summands,
+then  `sort_summands` sorts the summands on either side of the equality. -/
+meta def sort_summands (l : parse pexpr_list_or_texpr?) (locat : parse location) : tactic unit :=
+do
+  l : list expr ← (l.get_or_else []).mmap to_expr,
+  match locat with
+  | loc.wildcard := do
+    sort_summands_core tt l none,
+    ctx ← local_context,
+    ctx.mmap (λ e, sort_summands_core tt l (expr.local_pp_name e)),
+    skip
+  | loc.ns names := do
+    names.mmap $ sort_summands_core ff l,
+    skip
+  end
+
+end tactic.interactive
+
+open polynomial tactic
+open_locale polynomial classical
+
+variables {R : Type*} [semiring R] (f g h : R[X]) {r s t u : R} (r0 : t ≠ 0)
+
+example
+  (hp : (monomial 1) u + (g + (monomial 5) 1) + 5 * X + ((monomial 0) s + (monomial 2) t + f) +
+  (monomial 8) 1 = (3 * X + (monomial 8) 1 + (monomial 6) (t + 1)) + f + h + ((monomial 0) s +
+  (monomial 1) u) + (monomial 5) 1) :
+  (monomial 1) u + 5 * X + (g + (monomial 5) 1) + ((monomial 0) s + (monomial 2) t + f) +
+  (monomial 8) 1 = (3 * X + (monomial 8) 1 + (monomial 6) (t + 1)) + f + h + ((monomial 0) s +
+  (monomial 1) u) + (monomial 5) 1 :=
+begin
+--  convert hp using 1, ac_refl, -- works and takes 8s
+--  sort_summands at ⊢ hp, assumption, --takes under 600ms
+  sort_summands [g, (5 * X : R[X]), g, 3, f] at ⊢ hp,
+  sort_summands [(5 * X : R[X]), monomial 2 t, monomial 0 s],
+  sort_summands at ⊢ hp,
+  assumption,
+end
+
+example (hp : f = g) :
+  7 + f + (monomial 3 r + 42) + X ^ 5 * h = monomial 3 r + h * X ^ 5 + g + 7 + 42 :=
+begin
+  sort_summands [f, g],
+  congr' 3, -- takes care of using assumption `hp`
+  exact X_pow_mul,
+end
```

Comment:
These should be extracted to a separate test file [here](https://github.com/leanprover-community/mathlib/tree/master/test)

Reply:
Done